### PR TITLE
Add feature to enable upload of files to the rest api.

### PR DESCRIPTION
**Pull request by tcezard**
This PR adds ability to upload file through POST PATCH or PUT via rest_communication.
it expect to have the file name in the json and will detect the file and upload it using request multipart form.
the usage would be:
```python
json_list = [
    {'element_id': 'e1', 'report_r1': ('file', 'path/to/file/to/upload1.html')},
    {'element_id': 'e2', 'report_r1': ('file', 'path/to/file/to/upload2.html')},
] 
rest_communication.post_or_patch('endpoint', json_list, 'element_id')
```

### CODE REVIEW

**Review by katieemelianova — APPROVED** [review]

File: egcg_core/rest_communication.py [hunk 1]
```diff
@@ -100,12 +161,14 @@ def get_document(self, endpoint, idx=0, **query_args):
...
+        files, payload = self._detect_files_in_json(payload)
+        return self._req('POST', self.api_url(endpoint), json=payload, files=files)
 
     def put_entry(self, endpoint, element_id, payload):
-        return self._req('PUT', urljoin(self.api_url(endpoint), element_id), json=payload)
+        files, payload = self._detect_files_in_json(payload)
+        return self._req('PUT', urljoin(self.api_url(endpoint), element_id), json=payload, files=files)
 
-    def _patch_entry(self, endpoint, doc, payload, update_lists=None):
+    def _patch_entry(self, endpoint, doc, payload, update_lists=None, files_content=None):
```
Comment (mwhamgenomics) [inline egcg_core/rest_communication.py]: `files_content` here is never used

File: egcg_core/rest_communication.py [hunk 2]
```diff
@@ -20,6 +22,58 @@ def serialise(queries):
...
+                    json_dict[k][1],                                 # file name
+                    Communicator.get_file_content(json_dict[k][1]),  # file content
+                    mimetypes.guess_type(json_dict[k][1])[0]         # file mime type
+                )
+        if files:
+            # remove the extracted files to the original json
+            for k in files:
+                json_dict.pop(k)
+            return files
+        return None
```
Comment (mwhamgenomics) [inline egcg_core/rest_communication.py:60]: If we return `None` here, could we end up trying to push `{'fastqc_report': null}` to the Rest API? If so, we should either make that nullable or return an empty list here instead.

**Review by mwhamgenomics — CHANGES_REQUESTED** [review]

File: egcg_core/rest_communication.py (hunk 2 above)
Comment (tcezard) [reply egcg_core/rest_communication.py:60]: No `_extract_files` returns a dict of files per `json_dict` provided.
The purpose of returning `None` explicitly here is to indicate that no files were found in the `json_dict`

```python
json_dict = {'key1': 'val1'}
print(_extract_files(json_dict))
>>> None
json_dict = {'key1': 'val1', 'key2': ('file', '/path/to/file.txt')}
print(_extract_files(json_dict))
>>> {'key2': ('/path/to/file.txt', 'filecontent', 'plain/text')}
```

**Comment by tcezard** [comment]
I've added a test to check if we're uploading a nested json and files at the same time which raises an exception
Usage would be:
```python
#with simple json --> use only one query
json_list = [
    {'element_id': 'e1', 'key1': 'value1', 'report_r1': ('file', 'path/to/file/to/upload1.html')},
    {'element_id': 'e2', 'key1': 'value2', 'report_r1': ('file', 'path/to/file/to/upload2.html')}
] 
rest_communication.post_or_patch('endpoint', json_list, 'element_id')

#with nested json --> split in two queries
json_list = [
    {'element_id': 'e1', 'key1': {'k': 'v1'}},
    {'element_id': 'e2', 'key1': {'k': 'v2'}}
]
rest_communication.post_or_patch('endpoint', json_list, 'element_id')
json_list = [
    {'element_id': 'e1', 'report_r1': ('file', 'path/to/file/to/upload1.html')},
    {'element_id': 'e2', 'report_r1': ('file', 'path/to/file/to/upload2.html')}
]
rest_communication.post_or_patch('endpoint', json_list, 'element_id')

#with nested json in one query -->  raise RestCommunicationError
json_list = [
    {'element_id': 'e1', 'key1': {'k': 'v1'}, 'report_r1': ('file', 'path/to/file/to/upload1.html')},
    {'element_id': 'e2', 'key1': {'k': 'v2'}, 'report_r1': ('file', 'path/to/file/to/upload2.html')}
]
rest_communication.post_or_patch('endpoint', json_list, 'element_id')
raise  RestCommunicationError('Cannot upload files and nested json in one query')

```
I tried to make `rest_communication` run the two queries but could not get it to work for function such as `post` or `put` because they do not receive `element_id` which is require for any subsequent patch.

**Review by mwhamgenomics — APPROVED** [review]

**Pull request merged**